### PR TITLE
Added a sequential queue

### PR DESCRIPTION
**Pull request by mjp41**
This changes the slab lists to use a sequential queue.
They were previously stored in a stack.

This commit also tidies up some incomplete refactoring from the
initial snmalloc2 work.

### CODE REVIEW

File: src/ds/seqqueue.h
Comment (nwf) [inline src/ds/seqqueue.h:24]: If this could be templatised on the pointer types involved, then `FreeListBuilder` could be a collection of these even with the `FreeObject` shenanigans of #389 (if we also renamed `FreeObject::T::next_object` to `FreeObject::T::next`.)

```suggestion
  template<typename T,
    template<typename> typename HPtr = Pointer,
    template<typename> typename QPtr = Pointer>
  class SeqQueue
  {
    static_assert(
      std::is_same<decltype(T::next), QPtr<T>>::value, "T->next must be a queue pointer to T");

    HPtr<T> head{nullptr};
    QPtr<T>* end{reinterpret_cast<QPtr<T>*>(&head)};
```

**Review by nwf — APPROVED** [review]
LGTM as is, if you don't like the additional templating suggestion.

File: src/ds/seqqueue.h
Comment (mjp41) [reply src/ds/seqqueue.h]: To summarise, I'm happy in principle, but we need to make the `reinterpret_cast` such that the client know about it, and has to say it is okay.  This could be either by explicitly restricting the types on `HPtr` and `QPtr`, or by requiring a lambda that does the cast.

So going to leave this to your future PR.

**Pull request merged**
